### PR TITLE
Install kernel driver as kernel version independent package

### PR DESCRIPTION
**Pull request by Araneidae**
This pull request adds support for building an RPM package for installing the kernel driver as a kernel version independent package using [DKMS](https://wiki.archlinux.org/index.php/Dynamic_Kernel_Module_Support).

### CODE REVIEW

**Comment by mdavidsaver** [comment]
This looks useful.  A README would be helpful to give others a starting point for how to make use of these changes.  Doesn't need to be exhaustive, but at least an example invocation.

The only part which gives me pause is the udev rule `SUBSYSTEM=="uio", MODE="0666"`.  As default permissions go, this is a bit promiscuous.  At minimum it needs to be restricted to this kernel module (cf. mrmShared/linux/README).  Does Fedora/RHEL have any groups which might be suitable?  eg. Debian has `plugdev` (for [removable disks](https://wiki.debian.org/SystemGroups)) which I might "abuse".

```udev
KERNEL=="uio*", ATTR{name}=="mrf-pci", GROUP="plugdev", MODE="0660"
```


Alternately, omit this file and leave permissions to a sysadmin.

**Comment by Araneidae** [comment]
I also have reservations about the promiscuous permissions.  We do have a suitable group, but it's not a standard one: `dcs` (for Diamond ControlS).  I could make the group configurable ... but then where am I going to put the configuration file to read?  We don't have a top level CONFIG file for this kind of setting.

Didn't know about the `ATTR{name}` attribute (been a while since I wrote udev rules), that's definitely an improvement, will do that and repush.  Will probably add the configurable group name as well.

Where should the README go?  The instruction is:
1. Run `make` in the `dkms` directory
2. Install the RPM on your target system

I'd love build recipes for other targets (DEB in particular), but I don't know how to build those.

**Comment by mdavidsaver** [comment]
`ATTR{name}` may be old.  These matches vary with kernel and udev versions.  This is part of the reason I've only provided an example in mrmShared/linux/README so far.  FYI. The recipe I use when formulating udev rules:

```sh
udevdump () 
{ 
    [ -z "$1" -o ! -e "$1" ] && echo "udevdump </dev/somefile>" && return 0;
    udevadm info -a -p $(udevadm info -q path -n "$1") --attribute-walk
}
```

Use like:

```sh
$ udevdump /dev/video0
```

Will print out all the possible matches for this device file.

**Comment by Araneidae** [comment]
Well, the first stanza of what I get is:
```
looking at device '/devices/pci0000:00/0000:00:01.1/0000:02:00.0/0000:03:08.0/0000:08:00.0/uio/uio0':
    KERNEL=="uio0"
    SUBSYSTEM=="uio"
    DRIVER==""
    ATTR{name}=="mrf-pci"
    ATTR{event}=="0"
    ATTR{version}==""
```
so think I'll go for `SUBSYSTEM=="uio", ATTR{name}=="mrf-pci"`.

**Comment by Araneidae** [comment]
Perhaps a more configurable way of defining the udev rules would help, but I hope this can be a sensible starting point.

File: mrmShared/linux/dkms/mrf.spec.in [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+install -m 0644 %{_curdir}/99-mrf.rules     %{buildroot}%{udevdir}
+
+%files
+%{dkmsdir}/uio_mrf.c
+%{dkmsdir}/jtag_mrf.c
+%{dkmsdir}/mrf.h
+%{dkmsdir}/Makefile
+%{dkmsdir}/Kbuild
+%{dkmsdir}/dkms.conf
+%{udevdir}/99-mrf.rules
```
[the remark this answers is not in the document]
Comment (mdavidsaver) [reply mrmShared/linux/dkms/mrf.spec.in:39]: Renaming this file to eg. `50-mrf.rules` would be friendly.  I believe the convention is that `99-*.rules` are seen as reserved for admins to override other rules.

**Review by mdavidsaver — COMMENTED** [review]
I think this is almost ready.

A couple of notes for future reference.

```sh
udevadm info --export-db
```

Is handy to see the effects of `/lib/udev/rules.d/*.rules`.

A rule with `TAG+=uaccess` would not be appropriate in this case since IOCs mostly run without a session (as defined by systemd).

**Comment by Araneidae** [comment]
I've renumbered the .rules file.  I'm not going to split the rpm at the moment, I imagine that can be a separate merge if it happens.  Anything else to do?

**Comment by mdavidsaver** [comment]
Fair enough, I guess we can wait until someone complains about the udev rule failing :)

**Pull request merged**